interfaces/prompting/requestrules/requestrules.go

```diff
@@ -452,36 +463,47 @@ func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) []prom
 	var conflicts []prompting_errors.RuleConflict
 
 	addVariant := func(index int, variant patterns.PatternVariant) {
+		variantStr := variant.String()
+		existingEntry, exists := permVariants.VariantEntries[variantStr]
+		if !exists {
+			newVariantEntries[variantStr] = variantEntry{
+				Variant: variant,
+				Outcome: rule.Outcome,
+				RuleIDs: map[prompting.IDType]bool{rule.ID: true},
+			}
+			return
+		}
 		newEntry := variantEntry{
 			Variant: variant,
-			RuleID:  rule.ID,
+			Outcome: rule.Outcome,
+			RuleIDs: make(map[prompting.IDType]bool, len(existingEntry.RuleIDs)+1),
 		}
-		variantStr := variant.String()
-		conflictingVariantEntry, exists := permVariants.VariantEntries[variantStr]
-		switch {
-		case !exists:
-			newVariantEntries[variantStr] = newEntry
-		case rdb.isRuleWithIDExpired(conflictingVariantEntry.RuleID, rule.Timestamp):
-			expiredRules[conflictingVariantEntry.RuleID] = true
-			newVariantEntries[variantStr] = newEntry
-		default:
-			// XXX: If we ever switch to adding variants directly to the real
-			// variant entries map instead of a new map, check that the
-			// "conflicting" entry does not belong to the same rule that we're
-			// in the process of adding, which is possible if the rule's path
-			// pattern can render to duplicate variants. Ignore self-conflicts.
-
-			// Exists and is not expired, so there's a conflict
+		newEntry.RuleIDs[rule.ID] = true
+		newVariantEntries[variantStr] = newEntry
+		for id := range existingEntry.RuleIDs {
+			if rdb.isRuleWithIDExpired(id, rule.Timestamp) {
+				// Don't preserve expired rules, and don't care if they conflict
+				expiredRules[id] = true
+				continue
+			}
+			if existingEntry.Outcome == rule.Outcome {
+				// Preserve non-expired rule which doesn't conflict
+				newEntry.RuleIDs[id] = true
```

Comment:
does `existingEntry.RuleIDs` need to be updated someplace as well?

Reply:
We replace `existingEntry` with `newEntry` in its entirety at the end if we don't hit any conflicts. By leaving `existingEntry` unchanged, we can roll back and leave everything unchanged by simply not replacing `existinglyEntry` with `newEntry`.